underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AUnderFileSystem.java

```diff
@@ -548,263 +258,143 @@ private boolean copy(String src, String dst) {
-   *
-   * @param child the key of the child
-   * @param parent the key of the parent
-   * @return the child key with the parent prefix removed, null if the parent prefix is invalid
-   */
-  private String getChildName(String child, String parent) {
-    if (child.startsWith(parent)) {
-      return child.substring(parent.length());
-    }
-    LOG.error("Attempted to get childname with an invalid parent argument. Parent: {} Child: {}",
-        parent, child);
-    return null;
+  @Override
+  protected OutputStream createObject(String key) throws IOException {
+    return new S3AOutputStream(mBucketName, key, mManager);
   }
 
-  /**
-   * Gets the metadata associated with a non-root key if it represents a folder. This method will
-   * return null if the key is not a folder. If the key exists as a prefix but does not have the
-   * folder dummy file, a folder dummy file will be created.
-   *
-   * @param key the key to get the metadata for
-   * @return the metadata of the folder, or null if the key does not represent a folder
-   */
-  private ObjectMetadata getFolderMetadata(String key) {
-    Preconditions.checkArgument(!isRoot(key));
-    String keyAsFolder = convertToFolderName(stripPrefixIfPresent(key));
-    ObjectMetadata meta = null;
+  @Override
+  protected boolean deleteObject(String key) throws IOException {
     try {
-      meta = mClient.getObjectMetadata(mBucketName, keyAsFolder);
-      // If no exception is thrown, the key exists as a folder
+      mClient.deleteObject(mBucketName, key);
     } catch (AmazonClientException e) {
-      // It is possible that the folder has not been encoded as a _$folder$ file
-      try {
-        String dir = stripPrefixIfPresent(key);
-        String dirPrefix = PathUtils.normalizePath(dir, PATH_SEPARATOR);
-        // Check if anything begins with <folder_path>/
-        ObjectListing objs = mClient.listObjects(mBucketName, dirPrefix);
-        // If there are, this is a folder and we can create the necessary metadata
-        if (objs.getObjectSummaries().size() > 0) {
-          mkdirsInternal(dir);
-          meta = mClient.getObjectMetadata(mBucketName, keyAsFolder);
-        }
-      } catch (AmazonClientException ace) {
-        return null;
-      }
+      LOG.error("Failed to delete {}", key, e);
+      return false;
     }
-    return meta;
+    return true;
   }
 
-  /**
-   * @param key the key to get the object details of
-   * @return {@link ObjectMetadata} of the key, or null if the key does not exist
-   */
-  private ObjectMetadata getObjectDetails(String key) {
-    // We try to get the metadata as a file and then a folder without checking isFolder to reduce
-    // the number of calls to S3.
+  @Override
+  protected String getFolderSuffix() {
+    return FOLDER_SUFFIX;
+  }
+
+  @Override
+  protected ObjectListingResult getObjectListing(String key, boolean recursive)
+      throws IOException {
+    String delimiter = recursive ? "" : PATH_SEPARATOR;
```

Comment:
Common prefixes which have not already been marked by Alluxio will not be fetched correctly in the recursive case?